src/Storages/MergeTree/MergeTreeData.h

```diff
@@ -1348,6 +1348,92 @@ class MergeTreeData : public IStorage, public WithMutableContext
         const MergeListEntry * merge_entry,
         std::shared_ptr<ProfileEvents::Counters::Snapshot> profile_counters);
 
+    class PartMutationBackoffPolicy : public WithContext
+    {
+        struct PartMutationInfo
+        {
+            size_t retry_count = 0ul;
+            Poco::Timestamp latest_fail_time{};
+            UInt64 mutation_failure_version = 0ul;
+
+            Poco::Timestamp getNextMinExecutionTime() const
+            {
+                return  latest_fail_time +  (1 << retry_count) * 1000ul;
+            }
+        };
+
+        using DataPartsWithRetryInfo = std::unordered_map<String, PartMutationInfo>;
+        DataPartsWithRetryInfo failed_mutation_parts;
+        size_t max_pospone_power;
+        mutable std::mutex parts_info_lock;
+
+    public:
+        explicit PartMutationBackoffPolicy(ContextPtr global_context_)
+            : WithContext(global_context_)
+        {
+            size_t max_pospone_time_ms =  global_context_->getMaxPostponeTimeForFailedMutations();
```

Comment:
Maybe we could store `context.getSettings()` instead of storing the context ptr itself? Except for the setting, we don't use the context